# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [3]:
%%time
%%sql

SELECT sum(sales_amount) as revenue FROM factsales

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.
CPU times: user 2.26 ms, sys: 317 µs, total: 2.58 ms
Wall time: 9.6 ms


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [5]:
%%time
%%sql

SELECT ct.country, sum(sales_amount) as revenue
FROM factsales as fct
JOIN dimcustomer as ct ON (fct.customer_key = ct.customer_key)
GROUP BY ct.country
limit 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 3.1 ms, sys: 0 ns, total: 3.1 ms
Wall time: 14.2 ms


[('Cambodia', Decimal('191.47')),
 ('Turkey', Decimal('1662.12')),
 ('Germany', Decimal('831.04')),
 ('Chad', Decimal('135.68')),
 ('Madagascar', Decimal('93.78')),
 ('France', Decimal('374.04')),
 ('Iraq', Decimal('111.73')),
 ('Colombia', Decimal('709.41')),
 ('Japan', Decimal('3471.74')),
 ('American Samoa', Decimal('71.80')),
 ('Virgin Islands, U.S.', Decimal('122.68')),
 ('Vietnam', Decimal('746.28')),
 ('Kuwait', Decimal('111.74')),
 ('Ecuador', Decimal('390.13')),
 ('Nepal', Decimal('116.78')),
 ('Cameroon', Decimal('200.46')),
 ('Saudi Arabia', Decimal('523.79')),
 ('Bangladesh', Decimal('402.05')),
 ('Lithuania', Decimal('73.76')),
 ('Netherlands', Decimal('586.66'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [12]:
%%time
%%sql

SELECT dt.month, sum(sales_amount) as revenue
FROM factsales as fct
JOIN dimdate as dt ON (fct.date_key = dt.date_key)
GROUP BY (dt.month)
ORDER BY revenue DESC
limit 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 2.86 ms, sys: 388 µs, total: 3.24 ms
Wall time: 13.4 ms


[(4, Decimal('28559.46')),
 (3, Decimal('23886.56')),
 (2, Decimal('9631.88')),
 (1, Decimal('4824.43')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [26]:
%%time
%%sql

SELECT dt.month, st.country, sum(sales_amount) as revenue
FROM factsales as fct
JOIN dimdate as dt ON (fct.date_key = dt.date_key)
JOIN dimstore as st ON (fct.store_key = st.store_key)
GROUP BY (month, country)
ORDER BY month, country ASC
limit 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
10 rows affected.
CPU times: user 3.89 ms, sys: 0 ns, total: 3.89 ms
Wall time: 17.1 ms


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [20]:
%%sql

SELECT dt.month, st.country, sum(sales_amount) as revenue
FROM factsales as fct
JOIN dimdate as dt ON (fct.date_key = dt.date_key)
JOIN dimstore as st ON (fct.store_key = st.store_key)
GROUP BY GROUPING SETS ((), month, country, (month, country))
limit 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>